File: android/app/src/main/java/com/rawthermal/app/printservice/EscPosEncoder.kt
package com.rawthermal.app.printservice

import java.io.ByteArrayOutputStream

/**
 * ESC/POS command encoder for thermal printers
 * Ported from TypeScript EscPosEncoder.ts
 */
class EscPosEncoder {
    private val buffer = ByteArrayOutputStream()

    companion object {
        // Control codes
        private const val ESC: Byte = 0x1B
        private const val GS: Byte = 0x1D
        private const val LF: Byte = 0x0A
        private const val NUL: Byte = 0x00

        // Alignment
        const val ALIGN_LEFT = 0
        const val ALIGN_CENTER = 1
        const val ALIGN_RIGHT = 2
    }

    /**
     * Initialize printer (reset to default state)
     */
    fun initialize(): EscPosEncoder {
        buffer.write(byteArrayOf(ESC, 0x40))
        return this
    }

    /**
     * Set text alignment
     */
    fun align(alignment: Int): EscPosEncoder {
        buffer.write(byteArrayOf(ESC, 0x61, alignment.toByte()))
        return this
    }

    /**
     * Set bold mode
     */
    fun bold(enabled: Boolean): EscPosEncoder {
        buffer.write(byteArrayOf(ESC, 0x45, if (enabled) 1 else 0))
        return this
    }

    /**
     * Set underline mode
     */
    fun underline(enabled: Boolean): EscPosEncoder {
        buffer.write(byteArrayOf(ESC, 0x2D, if (enabled) 1 else 0))
        return this
    }

    /**
     * Set double-height and double-width
     */
    fun setFontSize(doubleWidth: Boolean, doubleHeight: Boolean): EscPosEncoder {
        val mode = (if (doubleWidth) 0x20 else 0) or (if (doubleHeight) 0x10 else 0)
        buffer.write(byteArrayOf(GS, 0x21, mode.toByte()))
        return this
    }

    /**
     * Print text
     */
    fun text(content: String): EscPosEncoder {
        buffer.write(content.toByteArray(Charsets.UTF_8))
        return this
    }

    /**
     * Print text followed by newline
     */
    fun line(content: String): EscPosEncoder {
        text(content)
        newline()
        return this
    }

    /**
     * Print newline
     */
    fun newline(): EscPosEncoder {
        buffer.write(byteArrayOf(LF))
        return this
    }

    /**
     * Feed paper by specified number of lines
     */
    fun feed(lines: Int): EscPosEncoder {
        buffer.write(byteArrayOf(ESC, 0x64, minOf(lines, 255).toByte()))
        return this
    }

    /**
     * Print raster image
     * Uses GS v 0 command for raster bit image
     *
     * @param imageData 1-bit packed image data (8 pixels per byte, MSB first)
     * @param width Image width in pixels
     * @param height Image height in pixels
     */
    fun image(imageData: ByteArray, width: Int, height: Int): EscPosEncoder {
        // Calculate width in bytes (8 pixels per byte)
        val widthBytes = (width + 7) / 8

        // GS v 0 command parameters
        val xL = widthBytes and 0xFF
        val xH = (widthBytes shr 8) and 0xFF
        val yL = height and 0xFF
        val yH = (height shr 8) and 0xFF

        // GS v 0 mode xL xH yL yH [image data]
        // mode 0 = normal mode
        buffer.write(byteArrayOf(
            GS, 0x76, 0x30, 0x00,
            xL.toByte(), xH.toByte(), yL.toByte(), yH.toByte()
        ))
        buffer.write(imageData)

        return this
    }

    /**
     * Cut paper (full cut)
     */
    fun cut(): EscPosEncoder {
        buffer.write(byteArrayOf(GS, 0x56, 0x00))
        return this
    }

    /**
     * Partial cut paper
     */
    fun partialCut(): EscPosEncoder {
        buffer.write(byteArrayOf(GS, 0x56, 0x01))
        return this
    }

    /**
     * Open cash drawer
     */
    fun openCashDrawer(): EscPosEncoder {
        buffer.write(byteArrayOf(ESC, 0x70, 0x00, 0x19, 0x78))
        return this
    }

    /**
     * Beep
     */
    fun beep(times: Int = 1, duration: Int = 100): EscPosEncoder {
        val t = minOf(times, 9)
        val d = minOf(duration / 50, 9)
        buffer.write(byteArrayOf(ESC, 0x42, t.toByte(), d.toByte()))
        return this
    }

    /**
     * Set print density
     * @param density 0-7 (0=lightest, 7=darkest)
     */
    fun setDensity(density: Int): EscPosEncoder {
        val d = density.coerceIn(0, 7)
        buffer.write(byteArrayOf(GS, 0x7C, d.toByte()))
        return this
    }

    /**
     * Print horizontal rule
     */
    fun horizontalRule(char: Char = '-', width: Int = 32): EscPosEncoder {
        line(char.toString().repeat(width))
        return this
    }

    /**
     * Get encoded data as ByteArray
     */
    fun encode(): ByteArray = buffer.toByteArray()

    /**
     * Clear the buffer
     */
    fun clear(): EscPosEncoder {
        buffer.reset()
        return this
    }

    /**
     * Get current buffer size
     */
    fun size(): Int = buffer.size()
}

File: src/services/escpos/EscPosEncoder.ts
/**
 * ESC/POS Encoder for thermal printer commands
 * Reference: https://reference.epson-biz.com/modules/ref_escpos/index.php
 */

export type Alignment = 'left' | 'center' | 'right'
export type FontSize = 'normal' | 'double-height' | 'double-width' | 'double'
export type BarcodeType = 'UPC-A' | 'UPC-E' | 'EAN13' | 'EAN8' | 'CODE39' | 'ITF' | 'CODABAR' | 'CODE93' | 'CODE128'

// ESC/POS Command Constants
const ESC = 0x1B
const GS = 0x1D
const LF = 0x0A

export class EscPosEncoder {
  private buffer: number[] = []

  constructor(_encoding: string = 'utf-8') {
    // Encoding parameter reserved for future charset support
  }

  /**
   * Initialize the printer (ESC @)
   */
  initialize(): this {
    this.buffer.push(ESC, 0x40)
    return this
  }

  /**
   * Set text alignment (ESC a n)
   */
  align(alignment: Alignment): this {
    const alignMap: Record<Alignment, number> = {
      'left': 0,
      'center': 1,
      'right': 2
    }
    this.buffer.push(ESC, 0x61, alignMap[alignment])
    return this
  }

  /**
   * Enable/disable bold text (ESC E n)
   */
  bold(enabled: boolean = true): this {
    this.buffer.push(ESC, 0x45, enabled ? 1 : 0)
    return this
  }

  /**
   * Enable/disable underline (ESC - n)
   */
  underline(enabled: boolean = true): this {
    this.buffer.push(ESC, 0x2D, enabled ? 1 : 0)
    return this
  }

  /**
   * Set font size (GS ! n)
   */
  setFontSize(size: FontSize): this {
    const sizeMap: Record<FontSize, number> = {
      'normal': 0x00,
      'double-height': 0x01,
      'double-width': 0x10,
      'double': 0x11
    }
    this.buffer.push(GS, 0x21, sizeMap[size])
    return this
  }

  /**
   * Print text and convert to bytes
   */
  text(content: string): this {
    const encoder = new TextEncoder()
    const bytes = encoder.encode(content)
    this.buffer.push(...bytes)
    return this
  }

  /**
   * Add newline (LF)
   */
  newline(): this {
    this.buffer.push(LF)
    return this
  }

  /**
   * Print text with newline
   */
  line(content: string): this {
    return this.text(content).newline()
  }

  /**
   * Feed n lines (ESC d n)
   */
  feed(lines: number = 1): this {
    this.buffer.push(ESC, 0x64, Math.min(lines, 255))
    return this
  }

  /**
   * Cut paper - full cut (GS V 0)
   */
  cut(): this {
    this.buffer.push(GS, 0x56, 0x00)
    return this
  }

  /**
   * Cut paper - partial cut (GS V 1)
   */
  partialCut(): this {
    this.buffer.push(GS, 0x56, 0x01)
    return this
  }

  /**
   * Print raster image (GS v 0)
   * @param imageData - Binary image data (1-bit per pixel)
   * @param width - Image width in pixels
   * @param height - Image height in pixels
   */
  image(imageData: Uint8Array, width: number, height: number): this {
    const widthBytes = Math.ceil(width / 8)
    const xL = widthBytes & 0xFF
    const xH = (widthBytes >> 8) & 0xFF
    const yL = height & 0xFF
    const yH = (height >> 8) & 0xFF

    // GS v 0 mode xL xH yL yH [image data]
    this.buffer.push(GS, 0x76, 0x30, 0x00, xL, xH, yL, yH)
    this.buffer.push(...imageData)
    return this
  }

  /**
   * Print QR Code (GS ( k)
   */
  qrcode(content: string, size: number = 6): this {
    const data = new TextEncoder().encode(content)
    const pL = (data.length + 3) & 0xFF
    const pH = ((data.length + 3) >> 8) & 0xFF

    // Set QR code model
    this.buffer.push(GS, 0x28, 0x6B, 0x04, 0x00, 0x31, 0x41, 0x32, 0x00)

    // Set QR code size
    this.buffer.push(GS, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x43, Math.min(size, 16))

    // Set QR code error correction level (L)
    this.buffer.push(GS, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x45, 0x30)

    // Store QR code data
    this.buffer.push(GS, 0x28, 0x6B, pL, pH, 0x31, 0x50, 0x30, ...data)

    // Print QR code
    this.buffer.push(GS, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x51, 0x30)

    return this
  }

  /**
   * Print barcode (GS k)
   */
  barcode(content: string, type: BarcodeType = 'CODE128', height: number = 80): this {
    const typeMap: Record<BarcodeType, number> = {
      'UPC-A': 65,
      'UPC-E': 66,
      'EAN13': 67,
      'EAN8': 68,
      'CODE39': 69,
      'ITF': 70,
      'CODABAR': 71,
      'CODE93': 72,
      'CODE128': 73
    }

    const data = new TextEncoder().encode(content)

    // Set barcode height (GS h n)
    this.buffer.push(GS, 0x68, Math.min(height, 255))

    // Set barcode width (GS w n)
    this.buffer.push(GS, 0x77, 2) // 2 = medium width

    // Set HRI position (GS H n) - below barcode
    this.buffer.push(GS, 0x48, 2)

    // Print barcode (GS k m n d1...dn)
    this.buffer.push(GS, 0x6B, typeMap[type], data.length, ...data)

    return this
  }

  /**
   * Open cash drawer (ESC p m t1 t2)
   */
  openCashDrawer(pin: 0 | 1 = 0): this {
    this.buffer.push(ESC, 0x70, pin, 25, 250)
    return this
  }

  /**
   * Beep (ESC B n t) - if supported
   */
  beep(times: number = 1, duration: number = 3): this {
    this.buffer.push(ESC, 0x42, Math.min(times, 9), Math.min(duration, 9))
    return this
  }

  /**
   * Set print density (GS | n)
   */
  density(level: 'light' | 'normal' | 'dark'): this {
    const densityMap = {
      'light': 0xF8,
      'normal': 0x00,
      'dark': 0x08
    } as const
    // Note: This command varies by printer model
    this.buffer.push(GS, 0x7C, densityMap[level])
    return this
  }

  /**
   * Create a horizontal rule
   */
  horizontalRule(char: string = '-', width: number = 32): this {
    return this.line(char.repeat(width))
  }

  /**
   * Print a receipt header
   */
  header(title: string, subtitle?: string): this {
    this.align('center')
      .setFontSize('double')
      .bold(true)
      .line(title)
      .setFontSize('normal')
      .bold(false)

    if (subtitle) {
      this.line(subtitle)
    }

    return this.horizontalRule().align('left')
  }

  /**
   * Print a key-value pair
   */
  keyValue(key: string, value: string, width: number = 32): this {
    const spaces = width - key.length - value.length
    const padding = spaces > 0 ? ' '.repeat(spaces) : ' '
    return this.line(`${key}${padding}${value}`)
  }

  /**
   * Clear the buffer
   */
  clear(): this {
    this.buffer = []
    return this
  }

  /**
   * Get the buffer as Uint8Array
   */
  encode(): Uint8Array {
    return new Uint8Array(this.buffer)
  }

  /**
   * Get buffer length
   */
  get length(): number {
    return this.buffer.length
  }
}

// Factory function for convenience
export function createEncoder(encoding?: string): EscPosEncoder {
  return new EscPosEncoder(encoding)
}
